crates/autopilot/src/database/onchain_order_events.rs

```diff
@@ -0,0 +1,696 @@
+        let (custom_order_data, broadcasted_order_data, orders) =
+            self.extract_custom_and_general_order_data(events).await?;
+
+        let _timer = Metrics::get()
+            .database_queries
+            .with_label_values(&["append_ethflow_order_events"])
+            .start_timer();
+        let mut transaction = self.db.0.begin().await?;
+
+        database::onchain_broadcasted_orders::append(
+            &mut transaction,
+            broadcasted_order_data.as_slice(),
+        )
+        .await
+        .context("append_onchain_orders failed")?;
+
+        self.custom_onchain_data_parser
+            .append_custom_order_info_to_db(&mut transaction, custom_order_data)
+            .await
+            .context("append_custom_onchain_orders failed")?;
+
+        database::orders::insert_orders(&mut transaction, orders.as_slice())
+            .await
+            .context("insert_orders failed")?;
+
+        transaction.commit().await.context("commit")?;
+        Ok(())
+    }
+
+    async fn replace_events(
+        &mut self,
+        events: Vec<EthContractEvent<ContractEvent>>,
+        range: std::ops::RangeInclusive<shared::event_handling::BlockNumber>,
+    ) -> Result<()> {
+        let (custom_onchain_data, broadcasted_order_data, orders) =
+            self.extract_custom_and_general_order_data(events).await?;
+
+        let _timer = Metrics::get()
+            .database_queries
+            .with_label_values(&["replace_onchain_order_events"])
+            .start_timer();
+
+        let mut transaction = self.db.0.begin().await?;
+
+        database::onchain_broadcasted_orders::mark_as_reorged(
+            &mut transaction,
+            range.start().to_u64() as i64,
+        )
+        .await
+        .context("mark_onchain_order_events failed")?;
+
+        self.custom_onchain_data_parser
+            .append_custom_order_info_to_db(&mut transaction, custom_onchain_data)
+            .await
+            .context("append_custom_onchain_orders failed")?;
+
+        database::onchain_broadcasted_orders::append(
+            &mut transaction,
+            broadcasted_order_data.as_slice(),
+        )
+        .await
+        .context("append_onchain_orders failed")?;
+
+        database::orders::insert_orders(&mut transaction, orders.as_slice())
+            .await
+            .context("insert_orders failed")?;
+        transaction.commit().await.context("commit")?;
+        Ok(())
+    }
+}
+
+impl<T: Send + Sync + Clone, W: Send + Sync> OnchainOrderParser<'_, T, W> {
+    async fn extract_custom_and_general_order_data(
+        &self,
+        events: Vec<EthContractEvent<ContractEvent>>,
+    ) -> Result<(
+        Vec<W>,
+        Vec<(database::events::EventIndex, OnchainOrderPlacement)>,
+        Vec<Order>,
+    )> {
+        let custom_event_data = self
+            .custom_onchain_data_parser
+            .parse_custom_event_data(&events)?;
```

Comment:
I think there are two kind of errors in parse_custom_event_data:
1. Unexpected errors: like parsing of indexing failing which should never happen:
E.g here: https://github.com/cowprotocol/services/pull/537/files#diff-0c97385d21ce6242b76434bdf73b384b70c0d4a288d3f664773b94e59ac3b7b8R31

and expected errors:
Errors could be for example that the quote_id is not parsed correctly. I think its nice if these errors are encapsulated and handled inside of `parse_custom_event_data`, as I am doing it in the PR: 
https://github.com/cowprotocol/services/pull/537/files#diff-0c97385d21ce6242b76434bdf73b384b70c0d4a288d3f664773b94e59ac3b7b8R35

Hence, I do implement the dropping of already in `parse_custom_event_data`. I think its nicer if `parse_custom_event_data` is allowed to decide whether an error is a real error or whether it should drop it off.

Reply:
> Hence, I do implement the dropping of already in parse_custom_event_data.

I didn't notice this when reviewing this PR, I think it's important to add this information to a comment in the trait so that people who implement it know that it is expected to drop bad events on parsing errors and not error out.

I see also that the filtering applies for other normal errors like failing to parse signatures, so overall it sounds good to me.